Add templating for default values

MarcelMue opened the pull request.

anvddriesch reviewed (approved):
Looks good to me!
Did I get it correctly that this only works in case the defaulted fields are missing entirely? In this case I will try to get rid of them in `kubectl-gs`.
As for the `sshKey` I think we are still working on how ssh should be managed to begin with so maybe this will change. But for now it seems good enough.

MarcelMue commented:
> Looks good to me!
> Did I get it correctly that this only works in case the defaulted fields are missing entirely? In this case I will try to get rid of them in `kubectl-gs`.
> As for the `sshKey` I think we are still working on how ssh should be managed to begin with so maybe this will change. But for now it seems good enough.

I want to try this out and also talk to upstream. It is possible to also do on the empty string but it makes the policy a bit more complicated (which I do not love).

anvddriesch commented:
I was just thinking, maybe we should have a human readable list of the rules? When we add more and more defaults here it could get hard to have an overview of what is taken care of and what isn"t.

MarcelMue commented:
> I was just thinking, maybe we should have a human readable list of the rules? When we add more and more defaults here it could get hard to have an overview of what is taken care of and what isn"t.

I have now tried to make the test cases we check easier to read, added some docs in the code and some comments in the yaml description. What would be the best way to make it more human readable for you?
I want to add a cleaner README with development instructions next - just looking for input :)

MarcelMue commented:
I added some more reviewers to get feedback while I make more improvements.

stone-z commented on helm/policies-aws/values.yaml:
```diff
@@ -0,0 +1,7 @@
+region: "eu-west-1"
```
Has there been any broader discussion/decision about where defaults should be defined? I was thinking it would be beneficial to have them in a CR in the cluster so that they are also accessible to UI tools and other projects which might want to read them. I also think it will be important to be able to determine with which version of defaults a given cluster was created.

MarcelMue replied:
We talked about this before and the current thinking was to utilize github.com/giantswarm/config for the definition of default values. Making them accessible would then happen through a configmap ideally.

Not sure if this is how we are going to keep it but to get things moving a CM should be sufficient.

anvddriesch commented:
> I have now tried to make the test cases we check easier to read, added some docs in the code and some comments in the yaml description. What would be the best way to make it more human readable for you?
I want to add a cleaner README with development instructions next - just looking for input :)

The comments already work for me. :D Thank you. A README will be very helpful as well.

calvix reviewed (approved):
looks ok to me, but my knowledge around this is zero :p

MarcelMue commented:
> > I have now tried to make the test cases we check easier to read, added some docs in the code and some comments in the yaml description. What would be the best way to make it more human readable for you?
> > I want to add a cleaner README with development instructions next - just looking for input :)
> 
> The comments already work for me. :D Thank you. A README will be very helpful as well.

I added some README in this PR already:
https://github.com/giantswarm/kyverno-policies/pull/43

MarcelMue commented:
I will merge here and work on the config management.
